File: frontend/src/lib/types/context.ts
import type { NPCBrief, NPCReaction } from './npc';
import type { GuidelinesResponse } from './rp';

export interface ContextRequest {
  user_message: string;
  last_response?: string;
  include_npc_reactions?: boolean;
}
export interface ContextDocument {
  name: string;
  card_type: string;
  file_path: string;
  source: 'keyword' | 'semantic' | 'graph' | 'trigger' | 'always_load';
  relevance_score: number;
  content: string | null;
  summary: string | null;
  status: 'new' | 'updated';
}
export interface ContextReference {
  name: string;
  card_type: string;
  status: 'already_loaded';
  sent_at_turn: number;
}
export interface FlaggedNPC {
  character: string;
  importance: string | null;
  reason: string;
}
export interface SceneState {
  location: string | null;
  time_of_day: string | null;
  mood: string | null;
  in_story_timestamp: string | null;
}
export interface CharacterState {
  location: string | null;
  conditions: string[];
  emotional_state: string | null;
}
export interface ThreadAlert {
  thread_id: string;
  name: string;
  level: 'gentle' | 'moderate' | 'strong';
  counter: number;
  threshold: number;
  consequence: string;
}
export interface TriggeredNote {
  trigger_id: string;
  trigger_name: string;
  inject_type: 'context_note' | 'state_alert';
  content: string;
  priority: number;
  signals_matched: string[];
}
export interface CardGap {
  entity_name: string;
  seen_count: number;
  suggested_type: string | null;
}
export interface StalenessWarning {
  type: string;
  exchange: number;
  failed_at: string;
  stale_fields: string[];
}
export interface WritingConstraints {
  text: string;
  patterns_included: string[];
  task_context: string;
  token_count: number;
}
export interface ResolveRequest {
  scene_description: string;
  keywords?: string[];
  max_hops?: number;
  max_results?: number;
}
export interface ContextResponse {
  current_exchange: number;
  documents: ContextDocument[];
  references: ContextReference[];
  npc_briefs: NPCBrief[];
  npc_reactions: NPCReaction[];
  flagged_npcs: FlaggedNPC[];
  guidelines: GuidelinesResponse | null;
  scene_state: SceneState;
  character_states: Record<string, CharacterState>;
  thread_alerts: ThreadAlert[];
  triggered_notes: TriggeredNote[];
  card_gaps: CardGap[];
  warnings: StalenessWarning[];
  writing_constraints: WritingConstraints | null;
}

File: rp_engine/models/context.py
"""Pydantic models for the context engine pipeline."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel

from rp_engine.models.npc import NPCReaction
from rp_engine.models.rp import GuidelinesResponse

# ---------------------------------------------------------------------------
# API Request / Response
# ---------------------------------------------------------------------------


class ContextRequest(BaseModel):
    user_message: str
    last_response: str | None = None
    include_npc_reactions: bool = True


class ContextDocument(BaseModel):
    name: str
    card_type: str
    file_path: str
    source: Literal["keyword", "semantic", "graph", "trigger", "always_load"]
    relevance_score: float
    content: str | None = None
    summary: str | None = None
    status: Literal["new", "updated"]


class ContextReference(BaseModel):
    name: str
    card_type: str
    status: Literal["already_loaded"] = "already_loaded"
    sent_at_turn: int


class NPCBrief(BaseModel):
    character: str
    importance: str | None = None
    archetype: str | None = None
    secondary_archetype: str | None = None
    behavioral_modifiers: list[str] = []
    trust_score: int = 0
    trust_stage: str | None = None
    emotional_state: str | None = None
    conditions: list[str] = []
    behavioral_direction: str = ""
    scene_signals: list[str] = []


class FlaggedNPC(BaseModel):
    character: str
    importance: str | None = None
    reason: str


class SceneState(BaseModel):
    location: str | None = None
    time_of_day: str | None = None
    mood: str | None = None
    in_story_timestamp: str | None = None


class CharacterState(BaseModel):
    location: str | None = None
    conditions: list[str] = []
    emotional_state: str | None = None


class ThreadAlert(BaseModel):
    thread_id: str
    name: str
    level: Literal["gentle", "moderate", "strong"]
    counter: int
    threshold: int
    consequence: str


class TriggeredNote(BaseModel):
    trigger_id: str
    trigger_name: str
    inject_type: Literal["context_note", "state_alert"]
    content: str
    priority: int = 0
    signals_matched: list[str] = []


class CardGap(BaseModel):
    entity_name: str
    seen_count: int
    suggested_type: str | None = None


class StalenessWarning(BaseModel):
    type: str = "stale_analysis"
    exchange: int
    failed_at: str
    stale_fields: list[str] = []


class WritingConstraints(BaseModel):
    text: str
    patterns_included: list[str] = []
    task_context: str = ""
    token_count: int = 0


class ContextResponse(BaseModel):
    current_exchange: int
    documents: list[ContextDocument] = []
    references: list[ContextReference] = []
    npc_briefs: list[NPCBrief] = []
    npc_reactions: list[NPCReaction] = []
    flagged_npcs: list[FlaggedNPC] = []
    guidelines: GuidelinesResponse | None = None
    scene_state: SceneState = SceneState()
    character_states: dict[str, CharacterState] = {}
    thread_alerts: list[ThreadAlert] = []
    triggered_notes: list[TriggeredNote] = []
    card_gaps: list[CardGap] = []
    warnings: list[StalenessWarning] = []
    writing_constraints: WritingConstraints | None = None


# ---------------------------------------------------------------------------
# Internal Types (used between services, not in API response)
# ---------------------------------------------------------------------------


class MatchedEntity(BaseModel):
    entity_id: str
    match_source: Literal["alias", "keyword", "name"]
    match_term: str
    score: float  # name=1.0, alias=0.8, keyword=0.5


class DetectedNPC(BaseModel):
    entity_id: str
    name: str
    detection_reason: str


class ExtractionResult(BaseModel):
    matched_entities: list[MatchedEntity] = []
    active_npcs: list[DetectedNPC] = []
    referenced_npcs: list[DetectedNPC] = []
    detected_locations: list[str] = []


# ---------------------------------------------------------------------------
# Graph resolution request (for /api/context/resolve)
# ---------------------------------------------------------------------------


class ResolveRequest(BaseModel):
    scene_description: str
    keywords: list[str] = []
    max_hops: int = 2
    max_results: int = 15
